extruct/dublincore.py

```diff
@@ -0,0 +1,157 @@
+    'linguisticsystem': 'http://purl.org/dc/terms/LinguisticSystem',
+    'license': 'http://purl.org/dc/terms/license',
+    'licensedocument': 'http://purl.org/dc/terms/LicenseDocument',
+    'mediator': 'http://purl.org/dc/terms/mediator',
+    'medium': 'http://purl.org/dc/terms/medium',
+    'physicalresource': 'http://purl.org/dc/terms/PhysicalResource',
+    'physicalmedium': 'http://purl.org/dc/terms/PhysicalMedium',
+    'modified': 'http://purl.org/dc/terms/modified',
+    'provenance': 'http://purl.org/dc/terms/provenance',
+    'provenancestatement': 'http://purl.org/dc/terms/ProvenanceStatement',
+    'publisher': 'http://purl.org/dc/terms/publisher',
+    'references': 'http://purl.org/dc/terms/references',
+    'replaces': 'http://purl.org/dc/terms/replaces',
+    'requires': 'http://purl.org/dc/terms/requires',
+    'rightsholder': 'http://purl.org/dc/terms/rightsHolder',
+    'source': 'http://purl.org/dc/terms/source',
+    'spatial': 'http://purl.org/dc/terms/spatial',
+    'location': 'http://purl.org/dc/terms/Location',
+    'subject': 'http://purl.org/dc/terms/subject',
+    'tableofcontents': 'http://purl.org/dc/terms/tableOfContents',
+    'temporal': 'http://purl.org/dc/terms/temporal',
+    'periodoftime': 'http://purl.org/dc/terms/PeriodOfTime',
+    'type': 'http://purl.org/dc/terms/type',
+    'valid': 'http://purl.org/dc/terms/valid',
+}
+
+_URL_NAMESPACES = ['http://purl.org/dc/terms/', 'http://purl.org/dc/elements/1.1/']
+
+
+def get_lower_attrib(name):
+    # get attribute to compare against _DC_TERMS or _DC_ELEMENTS
+    return re.sub(r".*\.", "", name).lower()
+
+
+class DublinCoreExtractor(object):
+    """DublinCore extractor following extruct API."""
+
+    def extract(self, htmlstring, base_url=None, encoding='UTF-8'):
+        tree = parse_html(htmlstring, encoding=encoding)
+        return list(self.extract_items(tree, base_url=base_url))
+
+    def extract_items(self, document, base_url=None):
+        elements = []
+        terms = []
+
+        def attrib_to_dict(attribs):
+            # convert _attrib type to dict
+            node_dict = {}
+            for attrib, value in attribs.items():
+                node_dict.update({attrib: value})
+            return node_dict
+
+        def populate_results(node, main_attrib):
+            # fill list with DC Elements or DC Terms
+            node_attrib = node.attrib
+            if main_attrib not in node_attrib:
+                return
+
+            name = node.attrib[main_attrib]
+            lower_name = get_lower_attrib(name)
+            if lower_name in _DC_ELEMENTS:
+                node.attrib.update({'URI': _DC_ELEMENTS[lower_name]})
+                elements.append(attrib_to_dict(node.attrib))
+
+            elif lower_name in _DC_TERMS:
+                node.attrib.update({'URI': _DC_TERMS[lower_name]})
+                terms.append(attrib_to_dict(node.attrib))
+
+        namespaces_nodes = document.xpath('//link[contains(@rel,"schema")]')
```

Comment:
I think there is a problem with using such xpaths - if you pass a subtree to extract data from, this xpath will refer to the whole document again. But we have such issue for other extractors as well, so probably it should be fixed separately, if we want to support passing subtrees instead of the complete documents.

Reply:
Sure, I would like to work on that in a separate issue.